keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,2249 @@
+    - Testing: Are there any tests for failure mode? If not, describe why.
+-->
+
+TBD.
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+TBD.
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+## Alternatives
+
+<!--
+What other approaches did you consider, and why did you rule them out? These do
+not need to be as detailed as the proposal, but should include enough
+information to express the idea and why it was not acceptable.
+-->
+
+### Pod annotations
+
+Instead of updating CRI and Kubernetes API in lock-step, the API changes could
+be split into two phases. Similar to e.g. how seccomp support was added, Pod
+annotations could be used as an initial user interface before the Kubernetes
+API changes are merged.
+
+Specifically, annotations for specifying RDT and blockio class would be
+supported. These are the two types of QoS-class resources that already have
+basic support in the container runtimes.
+
+- `rdt.resources.kubernetes.io/default` for setting a Pod-level default RDT
+  class for all containers
+- `rdt.resources.kubernetes.io/container.<container-name>` for
+  container-specific RDT class settings
+  blockio class for all containers
+- `blockio.resources.kubernetes.io/default` for setting a Pod-level default
+  blockio class for all containers
+- `blockio.resources.kubernetes.io/container.<container-name>` for
+  container-specific blockio class settings
+
+#### Kubelet
+
+Kubelet would interpret the specific [pod annotations](#pod-annotations) and
+translate them into corresponding `QOSResources` data in the CRI
+ContainerConfig message at container creation time (CreateContainerRequest).
+Pod-level QoS-class would not supported at this point (via pod annotations).
+
+A feature gate QOSResources would enable kubelet to interpretthe specific pod
+annotations. If the feature gate is disabled the annotations would simply be
+ignored by kubelet.
+
+#### API server
+
+A validation check (core api validation) would added in the API server to
+reject changes to the QoS-class resource specific
+[pod annotations](#pod-annotations) after a Pod has been created. This would
+ensure that the annotations always reflect the actual assignment of QoS-class
+resources of a Pod. It also would serve as part of the UX to indicate the
+in-place updates of the resources via annotations is not supported.
```

Comment:
:warning: It's unusual for us to reject metadata changes after an object is created. We usually limit that kind of restriction to apply to actual API fields.

Reply:
I'll leave this to highlight the behavior to reviewers. This comment shouldn't by itself block a merge.